vim-engine/src/main/kotlin/com/maddyhome/idea/vim/api/VimMarkServiceBase.kt

```diff
@@ -0,0 +1,605 @@
+    return when {
+      markChar.isGlobalMark() -> if (caret.isPrimary) setGlobalMark(caret.editor, char, offset) else false
+      markChar.isLocalMark() -> setMark(caret, char, offset)
+      else -> false
+    }
+  }
+
+  override fun setVisualSelectionMarks(editor: VimEditor) {
+    if (!editor.inVisualMode) return
+    val selectionType = SelectionType.fromSubMode(editor.subMode)
+    editor.carets()
+      .forEach {
+        val start = editor.offsetToBufferPosition(it.vimSelectionStart)
+        val end = editor.offsetToBufferPosition(it.offset.point)
+        it.lastSelectionInfo = SelectionInfo(start, end, selectionType)
+      }
+  }
+
+  override fun setVisualSelectionMarks(caret: ImmutableVimCaret, range: TextRange) {
+    setMark(caret, SELECTION_START_MARK, range.startOffset)
+    setMark(caret, SELECTION_END_MARK, range.endOffset)
+  }
+
+  override fun getVisualSelectionMarks(caret: ImmutableVimCaret): TextRange? {
+    return getMarksRange(caret, SELECTION_START_MARK, SELECTION_END_MARK)
+  }
+
+  override fun setChangeMarks(caret: ImmutableVimCaret, range: TextRange) {
+    setMark(caret, CHANGE_START_MARK, range.startOffset)
+    setMark(caret, CHANGE_END_MARK, range.endOffset - 1)
+  }
+
+  override fun getChangeMarks(caret: ImmutableVimCaret): TextRange? {
+    return getMarksRange(caret, CHANGE_START_MARK, CHANGE_END_MARK)
+  }
+
+  private fun getMarksRange(caret: ImmutableVimCaret, startMark: Char, endMark: Char): TextRange? {
+    val editor = caret.editor
+    val startOffset = getLocalMark(caret, startMark)
+      ?.offset(editor) ?: return null
+    val endOffset = getLocalMark(caret, endMark)
+      ?.offset(editor) ?: return null
+    return TextRange(startOffset, endOffset + 1)
+  }
+
+  override fun removeMark(editor: VimEditor, char: Char) {
+    val markChar = if (char == '`') '\'' else char
+    if (markChar.isLocalMark()) {
+      editor.carets()
+        .forEach { removeLocalMark(it, char) }
+    }
+    if (markChar.isGlobalMark()) {
+      removeGlobalMark(char)
+    }
+  }
+
+  override fun removeLocalMark(caret: ImmutableVimCaret, char: Char) {
+    val markChar = if (char == '`') '\'' else char
+    if (!markChar.isLocalMark()) return
+
+    if (caret.isPrimary) {
+      if (markChar == SELECTION_START_MARK) {
+        val selectionInfo = caret.lastSelectionInfo
+        val startPosition = selectionInfo.start
+        if (startPosition != null) caret.lastSelectionInfo = SelectionInfo(null, selectionInfo.end, selectionInfo.type)
+      } else if (markChar == SELECTION_END_MARK) {
+        val selectionInfo = caret.lastSelectionInfo
+        val endPosition = selectionInfo.end
+        if (endPosition != null) caret.lastSelectionInfo = SelectionInfo(selectionInfo.start, null, selectionInfo.type)
+      } else {
+        val path = caret.editor.getPath() ?: return
+        getLocalMarks(path).remove(markChar)
+      }
+    } else {
+      caret.markStorage.removeMark(markChar, caret)
```

Comment:
Shouldn't we have the logic above about the SELECTION_START_MARK also inside of markStorage.removeMark?